keps/sig-api-machinery/3157-watch-list/README.md

```diff
@@ -660,8 +661,12 @@ We expect no non-infra related flakes in the last month as a GA graduation crite
 - The Feature is implemented behind `WatchList` feature flag
 - Initial e2e tests completed and enabled
 - Scalability/Performance tests confirm gains of this feature
+- Add support for watchlist to APF
+
+#### Beta
 - Metrics are added to the kube-apiserver (see the [monitoring-requirements](#monitoring-requirements) section for more details)
 - Implement `SendInitialEvents` for `watch` requests in the etcd storage implementation
+- The feature is enabled for kubelet 
```

Comment:
Strongly in favor of beta requiring actual usage.

Blast radius is good consideration.  If we discover a client problem, would we have to update all kubelets, potentially skewed by three version, across N nodes?  By contrast, if we were to use the streaming list in kube-controller-manager, would we only have to change a single binary's launch?

This seems like another good candidate for working out how handle client library feature-gate like things.  @p0lyn0mial might be worth sync-ing with @benluddy about such gating.  We discussed it briefly last week for binary storage.

Reply:
If we find a bug we would have to backport it to the most recent three minor releases anyway.
But you are right that it might be more work for people that would like to deliver a potential patch to all nodes.
I'm fine with KCM as well.

I would like to avoid specifying in this document how we are going to enable that feature for a component. Additionally, I would prefer not to develop a generic mechanism for enabling features in the client-go library, assuming such a mechanism doesn't already exist.

Reply:
> If we find a bug we would have to backport it to the most recent three minor releases anyway.

That is fixing, but users need to be able to mitigate in minutes, not wait a month for patch release.


> Blast radius is good consideration. If we discover a client problem, would we have to update all kubelets, potentially skewed by three version, across N nodes? By contrast, if we were to use the streaming list in kube-controller-manager, would we only have to change a single binary's launch?

Thinking more about that, I agree with the suggestion. Also from the managed-provider perspective, the control over control-plane is much higher than the control over nodes.

What I'm wondering about is whether:
(a) we want to enable it for the whole kube-controller-manager
(b) enable it for O(1) controllers within kube-controller-manager
(c) maybe instead use scheduler?

Regarding enablement itself, I would use a different approach here, namely:
(a) we add an option at the informer level whether watch-list should be used [default to false]
(b) we create a new FG [let's say we decide for scheduler, then make it a scheduler FG]
(c) that new FG controls whether scheduler is using watch-list 

It doesn't solve the generic enablement of client-side features, but it would allow us to use standard mechanism and start getting real world experience with a well-known mitigation mechanism.

@deads2k - WDYT? ^^

Reply:
> (b) enable it for O(1) controllers within kube-controller-manager

It looks like it might be a bit hard to enable this feature for a single controller or even a single informer.
The issue is that controllers (KCM) accept informers which are created from a factory.
Usually we pass options to the factory not to individual informers (unless we would like to expose a new method on informers). Options are usually applied to all informers.
So it looks like it might even be hard to enable this feature for a single informer.

We could try to find a controller that directly creates an informer (bypasses the factory) but even in that case we would have to change the constructor signature for all informers. 

When I was looking on how to enable this feature for kubelet I was planing to pass a new FG via a flag (already defined) which would enable the streaming for all informers.

Reply:
OK - that's fine. We're still left with decision whether we choose (a) or (c) then.

As for enablement, I still believe that a better option is:
- create a new FG [directly in Beta]
- use that FG to decide whether watch-list is enabled in binary X
- if needed as a mitigation, we can disable the feature by switching off that new FG

Reply:
> As for enablement, I still believe that a better option is:

I'm not sure how it differs from what I wrote above but I think we will sort it out during a review :)